test/tools/isoltest.cpp

```diff
@@ -0,0 +1,347 @@
+/*
+	This file is part of solidity.
+
+	solidity is free software: you can redistribute it and/or modify
+	it under the terms of the GNU General Public License as published by
+	the Free Software Foundation, either version 3 of the License, or
+	(at your option) any later version.
+
+	solidity is distributed in the hope that it will be useful,
+	but WITHOUT ANY WARRANTY; without even the implied warranty of
+	MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+	GNU General Public License for more details.
+
+	You should have received a copy of the GNU General Public License
+	along with solidity.  If not, see <http://www.gnu.org/licenses/>.
+*/
+
+#include <libdevcore/CommonIO.h>
+#include <test/libsolidity/AnalysisFramework.h>
+#include <test/libsolidity/SyntaxTest.h>
+
+#include <boost/algorithm/string/replace.hpp>
+#include <boost/filesystem.hpp>
+#include <boost/program_options.hpp>
+
+#include <cstdlib>
+#include <iostream>
+#include <fstream>
+#include <queue>
+
+using namespace dev;
+using namespace dev::solidity;
+using namespace dev::solidity::test;
+using namespace dev::solidity::test::formatting;
+using namespace std;
+namespace po = boost::program_options;
+namespace fs = boost::filesystem;
+
+struct SyntaxTestStats
+{
+	int successCount;
+	int runCount;
+	operator bool() const { return successCount == runCount; }
+};
+
+class SyntaxTestTool
+{
+public:
+	SyntaxTestTool(string const& _name, fs::path const& _path, bool _formatted):
+		m_formatted(_formatted), m_name(_name), m_path(_path)
+	{}
+
+	enum class Result
+	{
+		Success,
+		Failure,
+		ParserError,
+		InputOutputError
+	};
+
+	Result process();
+
+	static SyntaxTestStats processPath(
+		fs::path const& _basepath,
+		fs::path const& _path,
+		bool const _formatted
+	);
+
+	static string editor;
+private:
+	enum class Request
+	{
+		Skip,
+		Rerun,
+		Quit
+	};
+
+	Request handleResponse(bool const _parserError);
+
+	void printContract() const;
+
+	bool const m_formatted;
+	string const m_name;
+	fs::path const m_path;
+	unique_ptr<SyntaxTest> m_test;
+};
+
+string SyntaxTestTool::editor;
+
+void SyntaxTestTool::printContract() const
+{
+	stringstream stream(m_test->source());
+	string line;
+	FormattedScope scope(cout, m_formatted, {CYAN});
+	while (getline(stream, line))
+		cout << "    " << line << endl;
+	cout << endl;
+}
+
+SyntaxTestTool::Result SyntaxTestTool::process()
+{
+	bool success;
+	bool parserError = false;
+	std::stringstream outputMessages;
+
+	(FormattedScope(cout, m_formatted, {BOLD}) << m_name << ": ").flush();
+
+	try
+	{
+		m_test = unique_ptr<SyntaxTest>(new SyntaxTest(m_path.string()));
+	}
+	catch (std::exception const& _e)
+	{
+		FormattedScope(cout, m_formatted, {RED}) << "cannot read test: " << _e.what() << endl;
+		return Result::InputOutputError;
+	}
+
+	try
+	{
+		success = m_test->run(outputMessages, "  ", m_formatted);
+	}
+	catch (...)
```

Comment:
I think I asked that before, but under which circumstances can we have an exception here?

Reply:
Ah nevermind, I found it. Can you please add a comment?

Reply:
Hm, actually in the medium term, this tool should also be used for parser tests, in which case we need to use a different framework and thus would not get an exception here, but we can keep it like that for now.

Reply:
Yes - in the medium term it would be best to use/write/modify a framework that gracefully returns both parser and syntax errors, instead of catching exceptions here.